network/src/noise_wrapper/handshake.rs

```diff
@@ -0,0 +1,387 @@
+                    client_timestamp_u64
+                ),
+            ));
+        }
+
+        // receive the initiation message
+        let mut client_init_message = [0u8; noise::handshake_init_msg_len(0)];
+        socket.read_exact(&mut client_init_message).await?;
+
+        // parse it
+        let (their_public_key, handshake_state, _) = self
+            .0
+            .parse_client_init_message(&prologue, &client_init_message)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+
+        // make sure the public key is a validator before continuing (if we're in the validator network)
+        if let Some(trusted_peers) = trusted_peers {
+            let found = trusted_peers
+                .read()
+                .unwrap()
+                .iter()
+                .find(|(_peer_id, public_keys)| public_keys.identity_public_key == their_public_key)
+                .is_some();
+            if !found {
+                println!("DEBUG: 4");
+                // TODO: security logging (mimoo)
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "noise: validator client initiated connection with an unknown public key: {}",
+                        their_public_key
+                    ),
+                ));
+            }
+        }
+
+        // check the timestamp is not a replay
+        {
+            let timestamps = LAST_SEEN_CLIENT_TIMESTAMPS.lock().unwrap();
+            if let Some(timestamp) = timestamps.get(&their_public_key) {
+                // TODO: security logging the ip + blocking the ip? (mimoo)
+                if timestamp == &client_timestamp_u64 {
+                    println!("DEBUG: 5");
+                    return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!(
+                        "noise: client initiated connection with a timestamp already seen before: {}",
+                        client_timestamp_u64
+                    ),
+                ));
+                }
+            }
+        }
+
+        // construct and send the response
+        let mut rng = rand::rngs::OsRng;
+        let mut server_response = [0u8; noise::handshake_resp_msg_len(0)];
+        let session = self
+            .0
+            .respond_to_client(&mut rng, handshake_state, None, &mut server_response)
+            .map_err(|e| {
+                io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("noise: wrong message received {}", e),
+                )
+            })?;
+        socket.write_all(&server_response).await?;
+
+        // the connection succeeded, store the client timestamp for replay prevention
+        {
```

Comment:
why are we putting this into a namespace?

Reply:
see above